### PR TITLE
bug: long name envs now stored using hashing

### LINKED ISSUE #648: Envs with >255 chars do not get saved
**Bug Description**
If an env is named with > 255 chars the corresponding json is not saved. Thus when visdom terminates all data is lost for this run. Note that while visdom is still operational the data can be accessed.

**Reproduction Steps**
Enter steps to reproduce the behavior:
1. send env with filename > 255 chars
2. run some data
3. terminate visdom
4. no json file present

**Expected behavior**
Filename should be hashed to size < 255 chars and saved to json so that data is never lost.

### PR DESCRIPTION
**Pull request by vedansh-5**
**Fix data loss for environment names exceeding OS filename limits**

## Description
This PR safely handles environment names (`eid`) that exceed the maximum filename length permitted by the operating system (e.g., > 255 characters). 

Fixes #648 

Changes introduced:

* Added fallback: on `OSError`, save env as `hash_<sha_256>.json`.
* Store original name inside JSON (`"name"` field).
* Load hashed files eagerly on startup and restore original names.
* Hide `hash_` files from UI; display real names from state.
* Fixed path bug in `compare_envs` and `load_env`; added hashed file fallback.
* Ensure hashed files are deleted when environments are removed.

## Motivation and Context
Currently, if an environment is named with > 255 characters, the corresponding JSON is not saved due to an unhandled `OSError` during `open(..., 'w')`. When the Visdom server terminates, all data for that run is permanently lost. This PR ensures that data is never lost by gracefully hashing problematic filenames and retaining full functionality and recovery after server restarts.

## How Has This Been Tested?
- Ran a local instance of Visdom and artificially generated an environment name exceeding 260 characters (`env='a'*260`).
- Sent plot and text data to the environment; observed successful disk write to `hash_<MD5>.json`.
- Terminated the Visdom server and restarted it. Verified that the long environment name successfully loaded back into the UI dropdown and all plotted data persisted accurately.
- Verified that deleting the long environment from the UI correctly clears the hashed file from disk.
- Tested standard (short) environment names to ensure normal lazy-loading and file IO remains completely unaffected.
- Ran `demo.py` on a clean branch and this branch to ensure standard functionality continues exactly as expected.

## How to test this PR?

 - run visdom locally 
 - run the following commands
  ``` markdown
from visdom import Visdom
import numpy as np

v = Visdom()
long_env_name = 'a' * 260

v.text('Testing long environment name.', env=long_env_name)
v.line(Y=np.array([1, 2, 3]), env=long_env_name)
v.save([long_env_name])
```
- rerun the server and try loading the env file - it should load now.
 
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Handle environments whose names exceed filesystem filename limits by transparently persisting them under hashed filenames while preserving their original names in application state.

Bug Fixes:
- Prevent data loss when saving environments with names that are too long for the operating system's filename limits by falling back to hashed filenames and storing the original name in the file.
- Fix environment loading and comparison to correctly resolve both normal and hashed environment files, including correcting the JSON path used for lookups.
- Ensure deletion of environments also removes any corresponding hashed files from disk so state and storage stay in sync.

Enhancements:
- Exclude hashed environment files from the visible environment list while still loading their contents and restoring original names during server startup.

### CODE REVIEW

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@vedansh-5 Whenever u get some time please go through the PR again and address co pilot review if needed

File: py/visdom/utils/server_utils.py [hunk 1]
```diff
@@ -120,11 +121,27 @@ def serialize_env(state, eids, env_path=DEFAULT_ENV_PATH):
...
-            with open(env_path_file, "w") as fn:
-                if isinstance(state[env_id], LazyEnvData):
-                    fn.write(json.dumps(state[env_id]._raw_dict))
-                else:
-                    fn.write(json.dumps(state[env_id]))
+            try:
+                with open(env_path_file, "w") as fn:
+                    if isinstance(state[env_id], LazyEnvData):
+                        state[env_id].lazy_load_data()
+                        fn.write(json.dumps(state[env_id]._raw_dict))
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply py/visdom/utils/server_utils.py:133]: Thanks for noting the test coverage here. Because Cypress primarily handles frontend interactions and cannot easily simulate a backend Python server restart to verify the disk-load behavior for hashed files, I tested this rigorously locally. We can follow up with a dedicated test suite for environments in a separate testing/chore PR."

File: py/visdom/utils/server_utils.py [hunk 2]
```diff
@@ -120,11 +121,31 @@ def serialize_env(state, eids, env_path=DEFAULT_ENV_PATH):
...
+                    env_path, "hash_{0}.json".format(hashed_id)
+                )
+                with open(env_path_file, "w") as fn:
+                    if isinstance(state[env_id], LazyEnvData):
+                        state[env_id].lazy_load_data()
+                        data_to_save = copy.deepcopy(state[env_id]._raw_dict)
+                    else:
+                        data_to_save = copy.deepcopy(state[env_id])
+                    data_to_save["name"] = env_id
+                    fn.write(json.dumps(data_to_save))
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply py/visdom/utils/server_utils.py:153]: The scenario where both a regular file and a hashed fallback exist simultaneously (e.g., from users migrating paths/drives) is rare. Since the core goal of this PR is solving the data-loss bug on initial save/load limits, I'd prefer to address the double-file side-by-side cleanup edge case in a dedicated follow-up PR to keep this ticket's footprint minimal.

**Comment by vedansh-5** [comment]
@Manik-Khajuria-5 @tonypzy @Jayantparashar10 Please review this PR - I have added the how to test instructions in PR description.

**Review by Manik-Khajuria-5 — APPROVED** [review]
@vedansh-5 Nice work the logic for handling failure cases works properly and also in ui it works properly . Have tested it locally and also read whole changes done and verified it 

Image

File: py/visdom/server/handlers/socket_handlers.py [hunk 3]
```diff
@@ -122,7 +122,19 @@ def on_message(self, message):
                 del self.state[msg["eid"]]
                 if self.env_path is not None:
                     p = os.path.join(self.env_path, "{0}.json".format(msg["eid"]))
-                    os.remove(p)
+                    if os.path.exists(p):
+                        os.remove(p)
+                    else:
+                        import hashlib
```
Comment (rajnisht7) [inline py/visdom/server/handlers/socket_handlers.py]: add this import at top of the file along with other imports

File: py/visdom/server/app.py [hunk 4]
```diff
@@ -167,8 +167,11 @@ def load_state(self):
         for env_json in env_jsons:
             eid = env_json.replace(".json", "")
             env_path_file = os.path.join(env_path, env_json)
+            import re
```
Comment (rajnisht7) [inline py/visdom/server/app.py]: here too put this at the top of file

File: py/visdom/utils/server_utils.py [hunk 5]
```diff
@@ -229,7 +250,13 @@ def window(args):
 
 def gather_envs(state, env_path=DEFAULT_ENV_PATH):
     if env_path is not None:
-        items = [i.replace(".json", "") for i in os.listdir(env_path) if ".json" in i]
+        import re
```
Comment (rajnisht7) [inline py/visdom/utils/server_utils.py]: here too

File: py/visdom/utils/server_utils.py [hunk 6]
```diff
@@ -120,11 +121,31 @@ def serialize_env(state, eids, env_path=DEFAULT_ENV_PATH):
...
-                    fn.write(json.dumps(state[env_id]))
+            try:
+                with open(env_path_file, "w") as fn:
+                    if isinstance(state[env_id], LazyEnvData):
+                        state[env_id].lazy_load_data()
+                        fn.write(json.dumps(state[env_id]._raw_dict))
+                    else:
+                        fn.write(json.dumps(state[env_id]))
+            except OSError as e:
+                import errno
```
Comment (rajnisht7) [inline py/visdom/utils/server_utils.py]: put this import at top

**Comment by rajnisht7** [comment]
verified the fix and works fine

File: py/visdom/server/handlers/web_handlers.py [hunk 7]
```diff
@@ -422,7 +423,15 @@ def wrap_func(handler, args):
...
-                os.remove(p)
+                if os.path.exists(p):
+                    os.remove(p)
+                else:
+                    hashed_id = hashlib.sha256(eid.encode("utf-8")).hexdigest()
+                    p = os.path.join(
+                        handler.env_path, "hash_{0}.json".format(hashed_id)
+                    )
+                    if os.path.exists(p):
+                        os.remove(p)
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply py/visdom/server/handlers/web_handlers.py]: @vedansh-5 why not add a check (already mentioned with copilot reivew)

**Review by rajnisht7 — CHANGES_REQUESTED** [review]

File: py/visdom/server/handlers/web_handlers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vedansh-5) [reply py/visdom/server/handlers/web_handlers.py]: Hi @rajnisht7 , Thanks for your feedback.
I prefer to keep the if/else block as is because - hashing fallback is only triggered by OS level nametoolong exceptions. For both files to exist simultaneously a user would have to migrate their environment files from a restrictive file system to a permissible file system, modify the file and save it again. 

Furthermore if both files do happen to exist its highly likely the result of manual file tampering or backups created by developer. In such case we should delete the primary file and not sweep the directory to delete fallback files against the user's manual setup.

If a user ends up with both versions of the file and decides to use the UI functionality to delete, they can simply hit delete a second time if the fallback environment reappears. It's safer to avoid deleting multiple env files within a single server operation - also with the new batch delete functionality this operation would tidy up even more - auxiliary files can be deleted along the primary files in a single user operation.

Please share your views on this. Thanks!

[the remark this answers is not in the document]
Comment (rajnisht7) [reply py/visdom/server/handlers/web_handlers.py]: yup, its highly unlikely though and we can skip it for now, but it can be one of the edge cases

**Comment by vedansh-5** [comment]
@Jayantparashar10 @tonypzy @rajnisht7 Please review.

**Review by Jayantparashar10 — CHANGES_REQUESTED** [review]
The narrowing from except Exception to except `(OSError, ValueError)` is the right direction, but there's one gap `tornado.escape.json_decode` returns whatever the JSON contains if a `.json` file holds valid but non-dict JSON like `[]` `env_data.get()` raises `AttributeError` which is not caught by `(OSError, ValueError`) and will crash the server on startup. 

Add a type check right after decoding to close the gap:

```
env_data = tornado.escape.json_decode(fn.read())
   if not isinstance(env_data, dict):
       raise ValueError(f"Expected dict, got {type(env_data).__name__}")
```
 This converts it to a `ValueError` which your except already handles — corrupted file gets skipped and logged like any other bad file.

**Review by tonypzy — APPROVED** [review]
Run locally and after fix, the generated json is saved to `~/.visdom` by default.
Image

**Review by Jayantparashar10 — APPROVED** [review]
LGTM

**Review by ArnavBallinCode — APPROVED** [review]
LGTM

**Review by Saksham-Sirohi — APPROVED** [review]

**Review by rajnisht7 — APPROVED** [review]
